hal/armv7m/stm32/l4/timer.c

```diff
@@ -146,18 +147,43 @@ void timer_setAlarm(time_t us)
 	}
 
 	hal_spinlockSet(&timer_common.sp, &sc);
-	setval = (timer_getCnt() + (u32)ticks) & 0xffff;
+
+	/* Check if there's pending overflow to be handled
+	 * to not lose it should there be next when sleeping */
+	timerval[0] = timer_getCnt();
+	arrpend = (*(timer_common.lptim + lptim_isr) & 2) != 0 ? 1 : 0;
+	timerval[1] = timer_getCnt();
+
+	if (timerval[0] > timerval[1] || arrpend != 0) {
+		++timer_common.upper;
+		*(timer_common.lptim + lptim_icr) = 2;
+	}
+
+	if (ticks >= 0xffff) {
+		setval = (timerval[1] - 1) & 0xffff;
+	}
+	else {
+		setval = (timerval[1] + (u32)ticks) & 0xffff;
+
+		if (((setval - timerval[1]) & 0xffff) < hal_timerUs2Cyc(1000)) {
+			/* There is too much risk that interrupt will arrive before wfi
+			 * This can happen regardless of previous checks - can be
+			 * caused by a modulo operation few lines above */
+			setval = (timerval[1] + hal_timerUs2Cyc(1000)) & 0xffff;
```

Comment:
You can define value from `hal_timerUs2Cyc(1000)` as a const to not invoke function with the same argument two times.